build/build.go

```diff
@@ -807,33 +803,40 @@ func notSupported(d driver.Driver, f driver.Feature) error {
 	return errors.Errorf("%s feature is currently not supported for %s driver. Please switch to a different driver (eg. \"docker buildx create --use\")", f, d.Factory().Name())
 }
 
-type dockerLoadCallback func(name string) (io.WriteCloser, func(), error)
+type dockerLoadCallback func(name string) func(map[string]string) (io.WriteCloser, error)
 
-func newDockerLoader(ctx context.Context, d DockerAPI, name string, mw *progress.MultiWriter) (io.WriteCloser, func(), error) {
-	c, err := d.DockerAPI(name)
-	if err != nil {
-		return nil, nil, err
-	}
-
-	pr, pw := io.Pipe()
-	started := make(chan struct{})
-	w := &waitingWriter{
-		PipeWriter: pw,
-		f: func() {
-			resp, err := c.ImageLoad(ctx, pr, false)
-			if err != nil {
-				pr.CloseWithError(err)
-				return
+func newDockerLoader(ctx context.Context, d DockerAPI, name string, mw *progress.MultiWriter) func(map[string]string) (io.WriteCloser, error) {
+	return func(m map[string]string) (io.WriteCloser, error) {
+		done := make(chan struct{})
+		c, err := d.DockerAPI(name)
+		if err != nil {
+			return nil, err
+		}
+		pr, pw := io.Pipe()
+		started := make(chan struct{})
+		w := &waitingWriter{
+			PipeWriter: pw,
+			f: func() {
+				resp, err := c.ImageLoad(ctx, pr, false)
+				if err != nil {
+					pr.CloseWithError(err)
+					return
+				}
+				prog := mw.WithPrefix("", false)
+				close(started)
+				progress.FromReader(prog, "importing to docker", resp.Body)
+			},
+			started: started,
+		}
+		go func() {
+			select {
+			case <-ctx.Done():
+				pr.Close()
+			case <-done:
```

Comment:
this channel is never closed, so this is only waiting on ctx.Done, in which case you don't need the select, but maybe you meant to close the done channel.